### PR TITLE
Fixed #32144 -- Made makemessages remove temporary files when locale path doesn't exist.

### PR DESCRIPTION
**Pull request by boxed**
The error case "Unable to find a locale path to store translations..." didn't clean up its files.

https://code.djangoproject.com/ticket/32144#ticket

### CODE REVIEW

**Review by felixxm — COMMENTED** [review]
@boxed Thanks for this patch :+1: Tests are required. Also, wrapping the entire method in `try ... finally ...` doesn't seem necessary, I would prefer to call
```python
for build_file in build_files:
    build_file.cleanup()
```
before raising a `CommandError`, or wrapping only
https://github.com/django/django/blob/6014fd89b3391d72c2e35cb1c6caaed059a51b41/django/core/management/commands/makemessages.py#L565-L587

**Comment by boxed** [comment]
I thought of just adding it in the one missing place. But that's what previous developers have done, and they missed one place which is why this PR exists. The try-finally block ensures that anyone who adds a new raise won't have to step on this mine. That's the entire reason try-finally exists...

> Tests are required

Are you serious right now? There are no tests for the old code, and this is an improvement. Are you saying I need to make a test that sets up a filesystem, some files, runs this entire code path, and then checks for the non-existence of temporary files? This seems like a huge waste of CI resources as this will slow down the test suite for all eternity.

**Comment by felixxm** [comment]
> That's the entire reason try-finally exists...

Yes, but it make code less readable to wrap 150+ lines where an error can appear in the last 20, IMO. 

> Are you serious right now? 

Yes, it looks feasible. We need a test if we don't want to introduce a regression in the future. 

> Are you saying I need to make a test that sets up a filesystem, some files, runs this entire code path, and then checks for the non-existence of temporary files? This seems like a huge waste of CI resources as this will slow down the test suite for all eternity.

Not really, you don't need to do most of these things. Please check `i18n/test_extraction.py` for an inspiration.

**Comment by boxed** [comment]
> Yes, but it make code less readable to wrap 150+ lines where an error can appear in the last 20, IMO.

Maybe. This is what the code is actually doing though. If the indent is really a big problem, we can make a private `_process_locale_dir` that doesn't do the cleanup, and the public `process_locale_dir` has the try/finally and calls `_process_locale_dir`.

And you're wrong btw, the error can come as early as the very first line inside the for loop. Think of it like this: what if we create a temporary file, add it to the list, and then we start over with the for loop. Now there's an error at the very beginning, like maybe `self.stdout.write` raises an exception? Now you have temporary files on disk. 

> Not really, you don't need to do most of these things. Please check i18n/test_extraction.py for an inspiration.

Seems like that test does literally this :( 

Yea ok, I guess it's not a big extra cost to check the non-existance of a file here. Seems rather iffy though, as really that check should be on all the different code paths that are possible through makemessages, including future code paths not dreamt of yet... Are you satisfied with a check for just the code path I am fixing?

**Comment by felixxm** [comment]
> > Yes, but it make code less readable to wrap 150+ lines where an error can appear in the last 20, IMO.
> 
> Maybe. This is what the code is actually doing though. If the indent is really a big problem, we can make a private `_process_locale_dir` that doesn't do the cleanup, and the public `process_locale_dir` has the try/finally and calls `_process_locale_dir`.

It's not about indentation. `try ... finally` should be small and contain only code that can actually raise an exception, IMO. We don't need to consider theoretical and unsupported paths. 

> And you're wrong btw, the error can come as early as the very first line inside the for loop. Think of it like this: what if we create a temporary file, add it to the list, and then we start over with the for loop. Now there's an error at the very beginning, like maybe `self.stdout.write` raises an exception? Now you have temporary files on disk.
> 
> > Not really, you don't need to do most of these things. Please check i18n/test_extraction.py for an inspiration.
> 
> Seems like that test does literally this :(
> 
> Yea ok, I guess it's not a big extra cost to check the non-existance of a file here. Seems rather iffy though, as really that check should be on all the different code paths that are possible through makemessages, including future code paths not dreamt of yet... Are you satisfied with a check for just the code path I am fixing?

Yes, should be fine, but I also would be happy to review tests for _"...future code paths not dreamt of yet."_ :moon:  if you believe they will make new tests more bulletproof. Thanks for you understanding.

I have a gut feeling that you don't like code reviews. We're all here to make Django better.

**Comment by boxed** [comment]
> It's not about indentation. try ... finally should be small and contain only code that can actually raise an exception, IMO. 

You mean self.stdout.write can't raise? And all those string formatting operations can't raise? `build_file_class` can't raise? I think I can go on, but I stopped at this point. 

I also disagree with your statement in its entirety. try/except I would accept that it should be used for small isolated blocks, but this is try/finally, a totally different construct made for exactly the purpose I am using it for.

> We don't need to consider theoretical and unsupported paths.

Sure. Except those aren't covered by the existing code, or the the suggestion you made. Not to mention the dry violation, or the risk of the clean up being modified and then not being updated in all the places. Which is why we have try/finally in python.

> Yes, should be fine

Ok.

> but I also would be happy to review tests for "...future code paths not dreamt of yet." 🌔 if you believe they will make new tests more bulletproof. 

I can do that. With the try/finally block. Again, that's why it exists.

> I have a gut feeling that you don't like code reviews. We're all here to make Django better.

Code reviews are fine. It's just that I'm fixing the problem and you're suggesting I make all the same mistakes that caused the problem in the first place.

**Comment by jacobtylerwalls** [comment]
This fat block is about 1/3 of the function. And yes, with `else: return` it could leave temp files on disk theoretically:

https://github.com/django/django/blob/edd3c82df4a30a6c1c4dbe1d905dd89d65fbf1c6/django/core/management/commands/makemessages.py#L525-L556

but only if there's a poor refactoring in the future, since this would have raised before the temp files were created:
https://github.com/django/django/blob/edd3c82df4a30a6c1c4dbe1d905dd89d65fbf1c6/django/core/management/commands/makemessages.py#L323-L325

I hear @boxed wants to make future development completely safe, but I also hear @felixxm that wrapping the whole thing just to make future development safer brings readability downsides. My instinct was going to be to suggest a sweet spot in the middle, but as a practical matter that fat block in the middle of the function is so easy to skim I don't know that I would bother having separate try blocks above and below.

**Comment by boxed** [comment]
@jacobtylerwalls notice that I suggested the middle ground of introducing a private method without the try/finally and the public function just calls that

**Comment by boxed** [comment]
Ah. And also note that my proposed change isn't just safe against future refactoring. It's also safe against the myriad of other reason the code can exit without cleaning up the temp files. Like I already mentioned above.

**Comment by carltongibson** [comment]
TBH I agree with @felixxm. We have to maintain this code, and from experience I'd say that it's read more often than theoretical issues actually appear. Can we make the `try` smaller as per the initial review please. Thanks both.

**Comment by boxed** [comment]
What do you mean make the try smaller? How small? The entire reason this bug existed in the first place was because the code didn't use any of the two python methods you should be using for clean up: try/finally or with. You can't make the try smaller without also re-introducing this bug.

**Comment by boxed** [comment]
If you think the diff is hard to read look at this: https://github.com/django/django/pull/13609/files?w=1

**Comment by carltongibson** [comment]
https://github.com/django/django/pull/13609#pullrequestreview-516561055

**Comment by boxed** [comment]
@carltongibson I already said why that fix isn't sufficient. For example we know that `build_file.preprocess()` can raise. How do we know? Because there's already a try/except around it! So we KNOW that there's a temporary file leak there too. So the try/finally has to cover that too, which means it has to cover the entire for loop. Which is what my PR is doing.

**Comment by boxed** [comment]
Oh, and https://github.com/django/django/pull/13609/files#diff-f6dcf84eeecbbfbad63c316715f09e3eaf55784c75295e97a7fa57f424a52205R564 for example. The more you look, the more you find.

**Comment by pope1ni** [comment]
> For example we know that `build_file.preprocess()` can raise. How do we know? Because there's already a try/except around it!

I'm sorry, but it's hard to take this argument seriously. The code you are referring to is the following:

https://github.com/django/django/blob/edd3c82df4a30a6c1c4dbe1d905dd89d65fbf1c6/django/core/management/commands/makemessages.py#L514-L522

In this case, a specific exception is being raised and suppressed. You cannot conflate that to assume that all manner of other exceptions are going to be raised. It makes more sense to reason about the function itself:

https://github.com/django/django/blob/edd3c82df4a30a6c1c4dbe1d905dd89d65fbf1c6/django/core/management/commands/makemessages.py#L98-L115

In there, yes, I can see two potential problems. We can either fail to read or fail to write. On quick inspection I would imagine that failing to write would happen on the first file being built as the directory in the work path is the same. So then there would be nothing to clean up... As for reading I could imagine a problem maybe with permissions, but then you arguably have bigger problems than some left over temporary files and this is an external problem.

Please can we calm this down a little bit - the tone comes across a bit too aggressive. The fellows work hard to try and accommodate all of our improvements but they are also responsible for maintaining the project long after we've had our pull requests merged. We contributors may sometimes disagree - @carltongibson and I often have different points of view :wink:  -  but we need sometimes need to compromise or know when it's not worth fighting a battle over something.

If you can come up with a concrete, not contrived, examples where you can make this fail through other paths in the form of one or more unit tests, then you may be able to make your case stronger. @felixxm has said he would review such test cases.

Thank you for taking the time to work on this improvement to Django.

**Comment by boxed** [comment]
> In this case, a specific exception is being raised and suppressed. 

And we know this is a totally exhaustive list of all exceptions that can be raised how? This is a rhetorical question obviously. 

> If you can come up with a concrete, not contrived, examples where you can make this fail through other paths in the form of one or more unit tests

You mean in addition to the case I already identified? I am submitting this PR because I got these temporary files on disk multiple times already and it took me a while to figure out where they came from. This is not a "contrived" example. This happened to me multiple times. There is a python construct to do cleanup, in fact there are two: try/finally and with. We can use try/finally here or a with block with for example an ExitStack.

>  the tone comes across a bit too aggressive.

Yes I realize this. I am having to explain what try/finally is to django core developers it seems. This is an absurd situation.

> but they are also responsible for maintaining the project long after we've had our pull requests merged

This is *my* point though. I am arguing for using the python feature specifically designed to do cleanup, to do clean up. It will work now, and it will work in the future. It will work for the cases we can foresee, and it will work for the cases we haven't thought about. 

So far there's one and only one argument against the try/finally:

> try ... finally should be small and contain only code that can actually raise an exception, IMO. 

But that's not at all reasonable. That's now what try/finally is for. try/finally is very different from try/except. 

In addition I have also suggested that if you won't budge on this issue, we can change the method do just contain:

```python
build_files = []
try:
    self._process_locale_dir(self, locale_dir, files, build_file)
finally:
   for....
```

This would satisfy the "small and contain only code that can actually raise", and it would remove the extra indent (even though it is claimed that this is irrelevant). It would also minimize the git blame, if you are concerned about that. 

You have to be concerned with more than just the big try/finally block since you obviously are more bothered by that than by the fact that the current code has cleanup copy pasted into two places, still has cleanup holes (one of which I got bitten by, and more that I have not yet been bitten by), and is vulnerably to introducing even more bugs in the future.

Like `popen_wrapper` can raise CommandError too (https://github.com/django/django/blob/9386586f31b8a0bccf59a1bff647cd829d4e79aa/django/core/management/utils.py#L13) which was pointed out to me by someone. I have already pointed this out above. 

I am trying to fix the code with a simple change, using a standard python feature specifically made for this use case.

**Comment by pope1ni** [comment]
You opened ticket-32144 about files not being cleaned up when hitting the exception raised in this branch:

https://github.com/django/django/blob/4ac2d4fa42e1659f328c35b6b8d4761b3419c11a/django/core/management/commands/makemessages.py#L578-L583

The suggested fix is to perform the same cleanup as raised in this branch prior to the exception being raised:

https://github.com/django/django/blob/4ac2d4fa42e1659f328c35b6b8d4761b3419c11a/django/core/management/commands/makemessages.py#L566-L572

Making that fix will solve the problem.

> And we know this is a totally exhaustive list of all exceptions that can be raised how? This is a rhetorical question obviously. Look at what it calls! You have to now audit all of...

Did I say that it was an exhaustive list? Not that I recall.

> You mean in addition to the case I already identified? I am submitting this PR because I got these temporary files on disk multiple times already and it took me a while to figure out where they came from. This is not a "contrived" example. This happened to me multiple times.

Nobody is doubting your original issue as described. It has even been pointed out where to fix that case. The issue is that you are proposing umpteen **other** hypothetical points of failure without a concrete example that they could actually happen.

> Yes I realize this. I am having to explain what try/finally is to django core developers it seems. This is an absurd situation.

And this is now just patronising. I'm going to take a step back here as I've said my bit, but I suggest that you engage with the reviewers rather than attack them with vitriol. Please may I also refer you to the [code of conduct](https://www.djangoproject.com/conduct/) for your consideration.

**Comment by boxed** [comment]
> Making that fix will solve the problem.

Sure. It won't make the code good though. It will make the code longer, harder to read, and have all the maintainability problems we've already covered, plus the basic fact that it doesn't clearly fix the bug that files are sometimes not cleaned up.

> The issue is that you are proposing umpteen other hypothetical points of failure without a concrete example that they could actually happen.

Well this the point of try/finally: that you don't need to even consider any other cases. It makes it super easy to just read the code and then know it does the right thing without reading the code of every function called in the function. The fact is that I don't know if any other exceptions can happen, and it's quite unfeasible to audit the code to find out. Just look at `templatize`, it's huge and calls lots of functions itself. Cleaning up files is simple: just use try/finally. Auditing hundreds or thousands of lines of code to see if they raise an exception is hard. 

In fact, I just found such a case by reading `templatize`: https://github.com/django/django/blob/d6aff369ad33457ae2355b5b210faf1c4890ff35/django/utils/translation/template.py#L126 I an also reproduce it quite trivially and indeed when I did it created tons of temporary files on my disk that it didn't clean up:

```
? iommi/templates/iommi/base.html.py
? iommi/templates/iommi/form/actions.html.py
? iommi/templates/iommi/form/bootstrap/row_checkbox.html.py
? iommi/templates/iommi/form/bulma/field.html.py
? iommi/templates/iommi/form/choice.html.py
? iommi/templates/iommi/form/choice_select2.html.py
? iommi/templates/iommi/form/form.html.py
? iommi/templates/iommi/form/heading.html.py
? iommi/templates/iommi/form/image_row.html.py
? iommi/templates/iommi/form/radio.html.py
? iommi/templates/iommi/form/row_checkbox.html.py
? iommi/templates/iommi/form/semantic_ui/radio.html.py
? iommi/templates/iommi/form/semantic_ui/row_checkbox.html.py
? iommi/templates/iommi/query/form.html.py
? iommi/templates/iommi/query/form_toggle_script.html.py
? iommi/templates/iommi/table/bootstrap/paginator.html.py
? iommi/templates/iommi/table/bulma/paginator.html.py
? iommi/templates/iommi/table/header.html.py
? iommi/templates/iommi/table/js_select_all.html.py
? iommi/templates/iommi/table/paginator.html.py
? iommi/templates/iommi/table/semantic_ui/paginator.html.py
? iommi/templates/iommi/table/table.html.py
? iommi/templates/iommi/table/table_container.html.py
? iommi/templates/iommi/table/table_header_rows.html.py
```

Reproduction is to put `{% blocktrans %}{% if foo %}{% endif %}{% endblocktrans %}` in a .html file, but not the first one since crashes on the first file will not lead to any temporary files being created prior to the crash.

**Comment by boxed** [comment]
So, let's summarize:

- The entire code block DOES need to be wrapped because it can fail.
- If you don't like big try/finally blocks we can make a small such block and call another (private) function in the body of the try.
- We can also do the same private function with a `with ExitStack(...)` construct if you prefer. Personally I think try/finally is simpler.

I will gladly change this PR to any of these suggested changes, and add a test. Is this acceptable?

**Review by recepsirin — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
Strolled by because I got a notification as a watcher. I simply suggest opening a new ticket for the issue raised in https://github.com/django/django/pull/13609#issuecomment-718713695 and describing the motivation for wrapping the entire function there. The fellows stay right on top of ticket triage, which should unclog this a bit. Then this patch's scope will be clearer: it will either be for solving one accepted ticket or two accepted tickets depending on how the new ticket is received--simple as that. Cheers, Jacob

**Comment by boxed** [comment]
@jacobtylerwalls sorry I didn't understand that. Should I just change this PR to do the wrap of the private function? It's really only one problem here with one solution so I don't see the point of a new issue?

**Comment by jacobtylerwalls** [comment]
Sure, that would get this moving, at the very least. I understand your view of the root cause here, I'm just suggesting that since the original ticket was accepted before (i.e. without) any discussion of there being a deeper root cause, that temp files are being created without a finally block to catch every scenario, what got this bogged down was that not every thread participant had a clear view of whether there were, as one participant put it, non-contrived examples of additional problems. I think you demonstrated that, but well after the thread went quiet, so I was suggesting a second ticket as a faster way to get a view from the community/mergers before investing more time into this. No harm in having two issues representing different steps to reproduce that happen to be solvable with one code change, but all the same, no harm in running with your idea of updating this patch with your preferred approach and getting another review.

**Comment by boxed** [comment]
@jacobtylerwalls I've updated this PR to do the private method thing. I looked at the tests but it really feels contrived and silly to make a test for the failure case. It just seems like it's a waste of CO2.

**Comment by jacobtylerwalls** [comment]
Thanks for the updates. I doubt you'll find support for the view that this doesn't need tests; in fact, that will almost certainly keep this PR stalled (given that community/fellows use the ticket flags Needs Improvement and Needs Tests to determine what's ripe for review).

Continuing this discussion also has a carbon cost. @smithdc1 did some great work minimizing unnecessary runs of the entire test suite by implementing pre-commit hooks.

**Comment by boxed** [comment]
Yea ok... I noticed there's already a test for a similar scenario, but it's a bit broken. I've improved that test case to be more robust so it covers at least some of the failure cases I've uncovered in this discussion.

**Comment by boxed** [comment]
Hmm... and the JS tests failed now. I don't seem to have permission to retry it though, but it's quite clearly a flaky test :(

File: tests/i18n/commands/templates/template_0_with_no_error.tpl
Comment (jacobtylerwalls) [inline tests/i18n/commands/templates/template_0_with_no_error.tpl:4]: Minor typo: "lexicographically"

Thanks for adding a test!

Comment (boxed) [reply tests/i18n/commands/templates/template_0_with_no_error.tpl]: Fixed!

**Comment by boxed** [comment]
@jacobtylerwalls Fixed the spelling error, and the tests now pass (again there seems to be a flaky js tests).

**Comment by boxed** [comment]
Could someone take a look at this again? It's a trivial thing and all points above have been dealt with as far as I understand.

**Comment by jacobtylerwalls** [comment]
Will give it a spin against the patch review checklist soon, meantime--first thing needed is to squash the commits and reword to use the same message as the PR title. Thanks.

**Comment by boxed** [comment]
Done.

**Review by carltongibson — COMMENTED** [review]
Hi @boxed. Thanks for your input here. I'm going to close this in favour of 
#14578, which implements changes in line with the review comments.

**Pull request closed without merging**
